File: e_mart/e_mart/custom_scripts/sales_invoice/sales_invoice.js
// Copyright (c) 2025, efeone and contributors
// For license information, please see license.txt

frappe.ui.form.on('Sales Invoice', {
	onload(frm) {
		frm.fields_dict.buyback_items.grid.wrapper.on('change', function () {
			update_total_buyback_amount(frm);
			update_rounded_total(frm);
		});
		set_finance_filter(frm);
		attach_sales_expense_grid_events(frm);
		calculate_total_expense(frm);
		update_emi_amount(frm)
	},
	refresh: function (frm) {
		if (frm.doc.docstatus === 0) {
			update_outstanding_amount(frm);
			update_rounded_total(frm);
			update_emi_amount(frm)
		}
		if (!frm.is_new() && frm.doc.sales_type === "EMI") {
			frm.add_custom_button(__('Create Finance Invoice'), function () {
				frappe.call({
					method: "e_mart.e_mart.custom_scripts.sales_invoice.sales_invoice.create_finance_invoice",
					args: {
						sales_invoice_name: frm.doc.name
					},
					callback: function (r) {
						if (r.message) {
							frappe.set_route('Form', 'Finance Invoice', r.message);
						}
					}
				});
			});
            if (frm.doc.docstatus === 1 && frm.doc.down_payment && frm.doc.down_payment_amount && frm.doc.down_payment_paid == 0 ) {
                frm.add_custom_button(__('Down Payment'), function() {
                    frappe.model.open_mapped_doc({
                        method: "e_mart.e_mart.custom_scripts.sales_invoice.sales_invoice.make_down_payment_entry",
                        source_name: frm.doc.name
                    });
                }, __('Create'));
            }
        }
	    calculate_total_expense(frm);
	},
	sales_type(frm) {
		set_finance_filter(frm);
	},
	// Triggered when relevant fields change for EMI recalculation
	down_payment_amount: function(frm) {
		update_emi_amount(frm);
		generate_emi_schedule(frm); 
	},
	total: function(frm) {
		update_outstanding_amount(frm);
		update_rounded_total(frm);
		update_emi_amount(frm);
		generate_emi_schedule(frm);
	},
	total_taxes_and_charges(frm) {
		update_outstanding_amount(frm);
		update_rounded_total(frm);
	},

	buyback_amount(frm) {
		update_outstanding_amount(frm);
		update_rounded_total(frm);
	},

	is_buyback(frm) {
		update_outstanding_amount(frm);
		update_rounded_total(frm);
	},
	no_of_installment: function(frm) {
		generate_emi_schedule(frm);
	},
	emi_amount: function(frm) {
		generate_emi_schedule(frm);
	},
	emi_date: function(frm) {
		generate_emi_schedule(frm);
	},
	validate(frm) {
		update_emi_amount(frm);
		generate_emi_schedule(frm);
		calculate_total_expense(frm);
        update_profit_for_commission(frm);
	},
	calculate_totals(frm) {
		setTimeout(() => {
			update_outstanding_amount(frm);
			update_rounded_total(frm);
			calculate_total_expense(frm);
		}, 200);
	},
    total_expense(frm) {
		update_profit_for_commission(frm);
	}
});
// Buyback Item child table
frappe.ui.form.on('Buyback Item', {
	qty(frm, cdt, cdn) {
		calculate_row_amount(frm, cdt, cdn);
	},
	rate(frm, cdt, cdn) {
		calculate_row_amount(frm, cdt, cdn);
	},
	buyback_items_remove(frm) {
		update_total_buyback_amount(frm);
	}
});
// Sales Invoice Items child table
frappe.ui.form.on('Sales Invoice Item', {
	qty(frm) {
		frm.trigger("calculate_totals");
	},
	rate(frm) {
		frm.trigger("calculate_totals");
	},
	amount(frm) {
		frm.trigger("calculate_totals");
		calculate_total_expense(frm);
	},
	items_add(frm) {
		frm.trigger("calculate_totals");
	},
	items_remove(frm) {
		frm.trigger("calculate_totals");
	},
	item_code(frm, cdt, cdn) {
        let row = locals[cdt][cdn];
        if (row.item_code) {
            frappe.db.get_value("Item", row.item_code, "sales_expense_contribution")
            .then(r => {
                if (r.message) {
                    frappe.model.set_value(cdt, cdn, "sales_expense_contribution", r.message.sales_expense_contribution);
                    update_profit_for_commission(frm);
                    }
                });
            }
    },
	sales_expense_contribution(frm, cdt, cdn) {
        update_profit_for_commission(frm);
    }
});

// Taxes child table
frappe.ui.form.on('Sales Taxes and Charges', {
	tax_amount(frm) {
		frm.trigger("calculate_totals");
	},
	rate(frm) {
		frm.trigger("calculate_totals");
	},
	taxes_add(frm) {
		frm.trigger("calculate_totals");
	},
	taxes_remove(frm) {
		frm.trigger("calculate_totals");
	}
});

// Calculate Buyback Row Amount
function calculate_row_amount(frm, cdt, cdn) {
	let row = locals[cdt][cdn];
	row.amount = (flt(row.qty) || 0) * (flt(row.rate) || 0);
	frm.fields_dict.buyback_items.grid.refresh();
	update_total_buyback_amount(frm);
}

// Sum all row amounts from buyback_items_amount
function update_total_buyback_amount(frm) {
	let total = 0;
	(frm.doc.buyback_items || []).forEach(row => {
		total += flt(row.amount || 0);
	});
	frm.set_value('buyback_amount', total);
	update_outstanding_amount(frm);
	update_rounded_total(frm);
	update_grand_total(frm);
}

/*
* Calculate Outstanding Amount based on buyback amount
*/
function update_outstanding_amount(frm) {
	if (frm.doc.docstatus === 0) { 
		const total = flt(frm.doc.total || 0);
		const taxes = flt(frm.doc.total_taxes_and_charges || 0);
		const buyback = flt(frm.doc.buyback_amount || 0);

		let outstanding = total + taxes;
		if (frm.doc.is_buyback) {
			outstanding -= buyback;
		}
		frm.set_value('outstanding_amount', Math.max(outstanding, 0));
		update_grand_total(frm); 
		update_emi_amount(frm);
	}
}

/*
* Calculate Rounded Total based on buyback amount
*/
function update_rounded_total(frm) {
	const total = flt(frm.doc.total || 0);
	const taxes = flt(frm.doc.total_taxes_and_charges || 0);
	const buyback = flt(frm.doc.buyback_amount || 0);

	let grand_total = total + taxes;
	if (frm.doc.is_buyback) {
		grand_total -= buyback;
	}
	frm.set_value('rounded_total', Math.round(grand_total));
}

/*
* Calculate Grand Total based on buyback amount
*/
function update_grand_total(frm) {
	const total = flt(frm.doc.total || 0);
	const taxes = flt(frm.doc.total_taxes_and_charges || 0);
	const buyback = flt(frm.doc.buyback_amount || 0);

	let grand_total = total + taxes;
	if (frm.doc.is_buyback) {
		grand_total -= buyback;
	}
	frm.set_value('grand_total', grand_total);
}

// Filter customer field to show only providers when sales_type is EMI
// If selected customer is not a provider, clear it
function set_finance_filter(frm) {
	frm.set_query("mode_of_payment", () => {
		if (frm.doc.sales_type === "EMI") {
			return {
				filters: {
					is_finance: 1
				}
			};
		} else {
			return {};
		}
	});

	frm.set_query("emi_provider", () => {
		if (frm.doc.sales_type === "EMI") {
			return {
				filters: {
					is_provider: 1
				}
			};
		} else {
			return {};
		}
	});
}

/**
 * Calculates the EMI amount after deducting the down payment
 * and updates the emi_amount field immediately.
 */
function update_emi_amount(frm) {
    if (frm.doc.docstatus !== 0) {
		return;
	}
	const down_payment = flt(frm.doc.down_payment_amount || 0);
	const outstanding = flt(frm.doc.outstanding_amount || 0);
	const emi_amount = outstanding - down_payment;

	frm.set_value('emi_amount', Math.max(emi_amount, 0));
}


/**
 * Generates EMI Duration child table rows dynamically
 * based on customer emi_start_date, emi_amount, and no_of_installment.
 */
function generate_emi_schedule(frm) {
	let emi_date = frm.doc.emi_date;
	let no_of_installments = frm.doc.no_of_installment;
	let emi_amount = frm.doc.emi_amount;

	if (!emi_date || !no_of_installments || !emi_amount) {
		return;
	}

	frm.clear_table('emi_duration');

	let installment_amount = Number(emi_amount) / Number(no_of_installments);
	let last_date = null;

	for (let i = 0; i < Number(no_of_installments); i++) {
		let installment_date = frappe.datetime.add_months(emi_date, i);
		last_date = installment_date; // Track last date

		frm.add_child('emi_duration', {
			date: installment_date,
			amount: installment_amount
		});
	}

	frm.refresh_field('emi_duration');

	// Set closing_date to last EMI date
	if (last_date) {
		frm.set_value('closing_date', last_date);
	}
}

frappe.ui.form.on('Sales Expenses', {
	amount(frm, cdt, cdn) {
		calculate_total_expense(frm);
	},
	sales_expenses_add(frm) {
		calculate_total_expense(frm);
	},
	sales_expenses_remove(frm) {
		calculate_total_expense(frm);
	}
});

/**
 * Attach grid-remove-row event to Sales Expenses child table.
 */
function attach_sales_expense_grid_events(frm) {
	if (frm._sales_expense_grid_attached) return;

	frm.fields_dict['sales_expenses'].grid.wrapper.on('grid-remove-row', () => {
		calculate_total_expense(frm);
	});

	frm._sales_expense_grid_attached = true;
}

/**
 * Sum up all amounts in Sales Expenses child table
 * and update total_expense field.
 */
function calculate_total_expense(frm) {
	let total = 0;

	(frm.doc.sales_expenses || []).forEach(row => {
		total += flt(row.amount || 0);
	});
	frm.set_value('total_expense', total);
	update_profit_for_commission(frm);
}

frappe.ui.form.on("Sales Team", {
	allocated_percentage: function(frm, cdt, cdn) {
			var sales_person = frappe.get_doc(cdt, cdn);
			let row = locals[cdt][cdn];
			if (sales_person.allocated_percentage) {
				sales_person.allocated_percentage = sales_person.allocated_percentage;
				sales_person.total_commission_rate = frm.doc.total_commission_rate;
				sales_person.incentive = flt(
					(sales_person.total_commission_rate * row.allocated_percentage) / 100.0,
				);
				frm.refresh_field('sales_team');
			}
	}
});

/**
 * Calculates the profit for Commission for each item by subtracting its Sales Expense Contribution
 * from the total expense.
 */
function update_profit_for_commission(frm) {
	const total_expense = flt(frm.doc.total_expense || 0);

	(frm.doc.items || []).forEach(row => {
		const contribution = flt(row.sales_expense_contribution || 0);
		const profit = total_expense - contribution;
		// Set profit_for_commission for each item
		frappe.model.set_value(row.doctype, row.name, 'profit_for_commission', profit);
	});

	frm.refresh_field('items');
}

File: e_mart/e_mart/custom_scripts/sales_invoice/sales_invoice.py
from datetime import datetime

import frappe
from frappe.model.mapper import get_mapped_doc
from frappe.utils import add_days, add_months, flt, get_first_day, get_last_day, get_url_to_form, nowdate


def on_submit(doc, method=None):
	create_scrap_stock_entry(doc, method)
	create_buyback_journal_entry(doc, method)
	create_demo_tasks_on_submit(doc, method)
	update_monthly_commission_log(doc, method)


def validate_buyback_fields(doc, method=None):
	"""
	Triggered on validation of Sales Invoice.

	1. Calculates amount for each Buyback Item row.
	2. Sums up all row amounts into buyback_amount.
	3. Adjusts the outstanding_amount, rounded total and grand total if is_buyback check box is checked
	"""

	total = 0

	# 1. Calculate amount per row
	if doc.buyback_items:
		for row in doc.buyback_items:
			row.amount = (row.qty or 0) * (row.rate or 0)
			total += row.amount

	# 2. Set total buyback amount
	doc.buyback_amount = total

	# 3.Calculate outstanding amount
	grand_total = (doc.total or 0) + (doc.total_taxes_and_charges or 0)
	if doc.is_buyback and doc.buyback_amount:
		grand_total -= doc.buyback_amount
	doc.outstanding_amount = round(grand_total)

	# 4. Set rounded_total and Grand Total
	grand_total = (doc.total or 0) + (doc.total_taxes_and_charges or 0)
	if doc.is_buyback and doc.buyback_amount:
		grand_total -= doc.buyback_amount
	doc.grand_total = grand_total
	doc.outstanding_amount = round(grand_total)
	doc.rounded_total = round(grand_total)


def create_scrap_stock_entry(doc, method):
	"""
	On submission of Sales Invoice:
	Creates a Stock Entry to move Buyback Items to Scrap Warehouse
	(Scrap warehouse is fetched from E-mart Settings).
	"""

	if not doc.buyback_items:
		return

	scrap_warehouse = frappe.db.get_single_value("E-mart Settings", "scrap_warehouse")
	if not scrap_warehouse:
		frappe.throw("Please set the Scrap Warehouse in E-mart Settings.")

	stock_entry = frappe.new_doc("Stock Entry")
	stock_entry.purpose = "Material Receipt"
	stock_entry.stock_entry_type = "Material Receipt"
	stock_entry.company = doc.company
	stock_entry.to_warehouse = scrap_warehouse
	stock_entry.sales_invoice_no = doc.name  # Optional

	for row in doc.buyback_items:
		if row.item and flt(row.qty) > 0:
			stock_entry.append(
				"items",
				{
					"item_code": row.item,
					"qty": row.qty,
					"uom": "Nos",  # adjust as needed
					"t_warehouse": scrap_warehouse,
					"basic_rate": row.rate,
				},
			)

	stock_entry.insert()
	stock_entry.submit()
	frappe.msgprint(
		f' Scrap Stock Entry Created: <a href="{frappe.utils.get_url_to_form(stock_entry.doctype, stock_entry.name)}" target="_blank"><b>{stock_entry.name}</b></a>',
		alert=True,
		indicator="green",
	)


def update_emi_amount(doc, method):
	"""
	Generate the emi amount after deducting the down payment
	"""
	down_payment = flt(doc.down_payment_amount)
	outstanding = flt(doc.outstanding_amount)
	doc.emi_amount = outstanding - down_payment


def generate_emi_schedule(doc, method):
	"""
	Generate EMI Duration table rows based on:
	- doc.emi_date (Sales Invoice EMI start date)
	- no_of_installments
	- emi_amount
	"""
	if doc.sales_type != "EMI":
		return
	if doc.is_buyback:
		return

	if not doc.emi_date:
		frappe.throw("Please set the EMI Start Date in Sales Invoice.")

	if not doc.no_of_installment:
		frappe.throw("Please set No of Installments.")

	if not doc.emi_amount:
		frappe.throw("Please set EMI Amount.")

	doc.set("emi_duration", [])  # Clear existing table

	installment_amount = flt(doc.emi_amount) / int(doc.no_of_installment)

	for i in range(int(doc.no_of_installment)):
		installment_date = add_months(doc.emi_date, i)
		doc.append("emi_duration", {"date": installment_date, "amount": installment_amount})


@frappe.whitelist()
def create_finance_invoice(sales_invoice_name):
	"""
	Create Finance Invoice from Sales Invoice
	"""
	sales_invoice = frappe.get_doc("Sales Invoice", sales_invoice_name)

	finance_invoice = frappe.new_doc("Finance Invoice")
	finance_invoice.customer = sales_invoice.emi_provider
	finance_invoice.posting_date = frappe.utils.nowdate()
	finance_invoice.total_amount = sales_invoice.total
	finance_invoice.actual_customer = sales_invoice.customer
	finance_invoice.total_taxes_and_charges = sales_invoice.total_taxes_and_charges

	for item in sales_invoice.items:
		finance_invoice.append(
			"items",
			{"actual_item": item.item_code, "qty": item.qty, "rate": item.rate, "amount": item.amount},
		)

	for tax in sales_invoice.taxes:
		finance_invoice.append(
			"sales_taxes_and_charges",
			{
				"charge_type": tax.charge_type,
				"account_head": tax.account_head,
				"description": tax.description,
				"rate": tax.rate,
				"tax_amount": tax.tax_amount,
				"total": tax.total,
				"tax_amount_after_discount_amount": tax.tax_amount_after_discount_amount,
			},
		)

	finance_invoice.save(ignore_permissions=True)

	return finance_invoice.name


#
@frappe.whitelist()
def make_down_payment_entry(source_name, target_doc=None):
	def set_missing_values(source, target):
		"""
		Create a Payment Entry from a Sales Invoice for down payment.
		Replaces total payment with down payment amount and map hte details.

		"""
		down_payment = source.down_payment_amount
		target.payment_type = "Receive"
		target.party_type = "Customer"
		target.paid_amount = down_payment
		target.received_amount = down_payment

		paid_from = frappe.get_value("Company", source.company, "default_receivable_account")
		paid_to = frappe.get_value("Company", source.company, "default_bank_account")

		target.paid_from = paid_from
		target.paid_to = paid_to
		target.posting_date = frappe.utils.nowdate()

		paid_from_currency = frappe.get_value("Account", paid_from, "account_currency")
		paid_to_currency = frappe.get_value("Account", paid_to, "account_currency")

		target.paid_from_account_currency = paid_from_currency
		target.paid_to_account_currency = paid_to_currency

		target.append(
			"references",
			{
				"reference_doctype": "Sales Invoice",
				"reference_name": source.name,
				"total_amount": source.rounded_total,
				"outstanding_amount": source.outstanding_amount,
				"allocated_amount": down_payment,
			},
		)

	doc = get_mapped_doc(
		"Sales Invoice",
		source_name,
		{
			"Sales Invoice": {
				"doctype": "Payment Entry",
				"field_map": {"customer": "party", "customer_name": "party_name", "company": "company"},
			}
		},
		target_doc,
		set_missing_values,
	)
	return doc


def create_buyback_journal_entry(doc, method):
	"""
	Create Buyback Journal entry for Buyback amount from Sales invoice.
	"""
	if not doc.is_buyback or not doc.buyback_amount or doc.buyback_amount <= 0:
		return

	customer_account = doc.debit_to
	if not customer_account:
		frappe.throw(f"Customer receivable account (debit_to) not found in Sales Invoice {doc.name}")

	company = doc.company or frappe.db.get_value("Sales Invoice", doc.name, "company")
	if not company:
		frappe.throw(f"Company not found for Sales Invoice {doc.name}")

	buyback_account = frappe.db.get_single_value("E-mart Settings", "buyback_posting_account")
	if not buyback_account:
		frappe.throw("Please set 'Buyback Posting Account' in E-mart Settings.")

	# Create Journal Entry
	journal_entry = frappe.new_doc("Journal Entry")
	journal_entry.voucher_type = "Credit Note"
	journal_entry.posting_date = nowdate()
	journal_entry.company = company
	journal_entry.remark = f"Buyback adjustment for Sales Invoice {doc.name}"

	# Credit Customer (receivable)
	journal_entry.append(
		"accounts",
		{
			"account": customer_account,
			"party_type": "Customer",
			"party": doc.customer,
			"credit_in_account_currency": doc.buyback_amount,
			"reference_type": "Sales Invoice",
			"reference_name": doc.name,
		},
	)

	# Dedit Buyback Posting Account
	journal_entry.append(
		"accounts", {"account": buyback_account, "debit_in_account_currency": doc.buyback_amount}
	)

	journal_entry.insert(ignore_permissions=True)
	journal_entry.submit()

	# Optional: Save Journal Entry reference if field exists
	if frappe.get_meta(doc.doctype).has_field("buyback_journal_entry"):
		doc.db_set("buyback_journal_entry", journal_entry.name)

	frappe.msgprint(
		f'Buyback Journal Entry Created: <a href="{frappe.utils.get_url_to_form("Journal Entry", journal_entry.name)}" target="_blank"><b>{journal_entry.name}</b></a>',
		alert=True,
		indicator="green",
	)


def create_demo_tasks_on_submit(doc, method=None):
	"""
	Create a Task for each Sales Invoice Item where is_demo_reqd is checked.
	Subject, Task Type, Minimal Duration, Escalation Duration come from E-mart Settings.
	Maps invoice_date, invoice_reference, customer, exp_start_date, and end_date into Task.
	"""

	settings = frappe.get_single("E-mart Settings")
	task_subject_template = settings.subject or "Demo Required - {item_name} [{invoice_name}]"
	task_type = settings.task_type or "Demo"
	minimal_duration = settings.minimal_duration or 0
	escalation_duration = settings.escalation_duration or 0

	for item in doc.items:
		if item.get("is_demo_reqd"):
			subject = task_subject_template.format(item_name=item.item_name, invoice_name=doc.name)

			exp_start_date = add_days(doc.posting_date, minimal_duration)
			end_date = add_days(doc.posting_date, escalation_duration)

			task = frappe.new_doc("Task")
			task.subject = subject
			task.type = task_type
			task.status = "Open"
			task.reference_type = "Sales Invoice"
			task.reference_name = doc.name
			task.description = (
				f"Demo required for Item: {item.item_name} ({item.item_code})\n"
				f"Qty: {item.qty}\n"
				f"Customer: {doc.customer}\n"
			)

			task.invoice_date = doc.posting_date
			task.invoice_reference = doc.name
			task.customer = doc.customer
			task.exp_start_date = exp_start_date
			task.exp_end_date = end_date

			task.insert(ignore_permissions=True)


def update_monthly_commission_log(doc, method):
	"""
	Create or update Monthly Commission Log for each Sales Person in the Sales Invoice.
	Logs incentives and invoice details for the respective employee and month.
	"""
	invoice_date = datetime.strptime(doc.posting_date, "%Y-%m-%d").date()
	month_start = get_first_day(invoice_date)
	month_end = get_last_day(invoice_date)
	month_name = invoice_date.strftime("%B")

	for sales_team_member in doc.sales_team:
		sales_person = sales_team_member.sales_person
		incentives = sales_team_member.incentive or 0

		# Get the Employee linked to Sales Person
		employee = frappe.db.get_value("Sales Person", sales_person, "employee")
		if not employee:
			frappe.log_error(f"Sales Person {sales_person} has no linked Employee.", "Monthly Commission Log")
			continue

		# Check if Monthly Commission Log exists
		log_name = frappe.db.exists(
			"Monthly Commission Log",
			{"employee": employee, "log_month": month_name, "start_date": month_start, "end_date": month_end},
		)

		if log_name:
			log = frappe.get_doc("Monthly Commission Log", log_name)
		else:
			log = frappe.new_doc("Monthly Commission Log")
			log.employee = employee
			log.log_month = month_name
			log.start_date = month_start
			log.end_date = month_end

		# Append detail row
		log.append(
			"monthly_commission_log",
			{
				"sales_invoice": doc.name,
				"date": invoice_date,
				"total_amount": doc.base_grand_total,
				"incentives": incentives,
			},
		)

		log.save(ignore_permissions=True)
		link = get_url_to_form("Monthly Commission Log", log.name)
		frappe.msgprint(
			f'Monthly Commission Log Created/Updated: <a href="{link}" target="_blank"><b>{log.name}</b></a>',
			alert=True,
			indicator="green",
		)


def calculate_total_expense(doc, method):
	"""
	Calculate and set the total of all sales_expenses in the document.
	"""
	total = 0

	for row in doc.get("sales_expenses", []):
		total += flt(row.amount or 0)
	doc.total_expense = total


# calculate incentives based on commission rate and allocated percentage
def map_commission_to_sales_team(doc, method):
	"""Fetch total_commission_rate to each sales_team row as allocated_amount and compute incentives."""
	commission_rate = doc.total_commission_rate or 0

	for row in doc.sales_team:
		row.total_commission_rate = commission_rate
		allocated_percentage = row.allocated_percentage or 0
		row.incentive = round((commission_rate * allocated_percentage) / 100, 2)


def calculate_profit_for_commission(doc, method):
	"""Calculates profit for commission for each item by subtracting its Sales Expense Contribution from the total expense."""
	total_expense = flt(doc.total_expense or 0)

	for item in doc.items:
		contribution = flt(item.sales_expense_contribution or 0)
		profit = total_expense - contribution
		item.profit_for_commission = profit
